crates/typst-kit/src/download.rs

```diff
@@ -0,0 +1,264 @@
+// Acknowledgement:
+// Closely modelled after rustup's `DownloadTracker`.
+// https://github.com/rust-lang/rustup/blob/master/src/cli/download_tracker.rs
+
+//! Helpers for making various web requests. These are primarily use for
+//! communicating with package registries.
+
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::io;
+use std::io::{ErrorKind, Read};
+use std::path::PathBuf;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use ecow::EcoString;
+use native_tls::{Certificate, TlsConnector};
+use once_cell::sync::OnceCell;
+use ureq::Response;
+
+/// Manages progress reporting for downloads.
+pub trait Progress {
+    /// Invoked when a download is started.
+    fn print_start(&mut self);
+
+    /// Invoked repeatedly while a download is ongoing.
+    fn print_progress(&mut self, state: &DownloadState);
+
+    /// Invoked when a download is finished.
+    fn print_finish(&mut self, state: &DownloadState);
+}
+
+/// An implementation of [`Progress`] whth no-op reporting, i.e. reporting
+/// events are swallowed.
+pub struct ProgressSink;
+
+impl Progress for ProgressSink {
+    fn print_start(&mut self) {}
+    fn print_progress(&mut self, _state: &DownloadState) {}
+    fn print_finish(&mut self, _state: &DownloadState) {}
+}
+
+/// The current state of an in progress or finished download.
+#[derive(Debug)]
+pub struct DownloadState {
+    /// The expected amunt of bytes to download, `None` if the response header
+    /// was not set.
+    pub content_len: Option<usize>,
+    /// The total amount of downloaded bytes until now.
+    pub total_downloaded: usize,
+    /// A backlog of the amount of downloaded bytes each second.
+    pub bytes_per_second: VecDeque<usize>,
+    /// The download starting instant.
+    pub start_time: Instant,
+}
+
+/// A minimal https client for downloading various resources.
+pub struct Downloader {
+    user_agent: EcoString,
+    cert_path: Option<PathBuf>,
+    cert: OnceCell<Option<Certificate>>,
+}
+
+impl Debug for Downloader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // NOTE: Certificate does not implement Debug, likely for security
+        //       reasons, but we still want to show a field that indicates
+        //       whether a certificate was loaded and is in use, so we implement
+        //       a small helper because `.field_with()` is not yet stable
+        // TODO: remove this once `.field_with` is stabilized (feature =
+        //       debug_closure_helpers)
+        struct CertExists;
+
+        impl Debug for CertExists {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(f, "<Certificate>")
```

Comment:
`Certificate(..)` is more common in the code base. But honestly I think it could simply be `Downloader(..)`.

Reply:
I'd go with `Certificate(..)` because I think being able to see if a certificate was loaded will aid in debugging, considering any IO error from loading it is swallowed.
Could perhaps also store the result and simply do `.ok()` in the getter, but I don't know if this interferes with any auto traits.

Reply:
Well that question has seemingly answered itself, on main this was already expanded to include the Result, so I'll go with that.